### PR TITLE
Declaration variables highlights

### LINKED ISSUE #85: Parameterized types do not highlight variable name
`List list` works fine, correctly highlighting list, however `List<Map> list` doesn't highlight list at all.

screenshot_20170809_202712

### PR DESCRIPTION
**Pull request by vincekd**
replacement for PR https://github.com/Groovy-Emacs-Modes/groovy-emacs-modes/pull/88 that (hopefully) addresses issues raised there.

[fix #85]

### CODE REVIEW

File: groovy-mode.el [hunk 1]
```diff
@@ -334,7 +296,178 @@ The function name is the second group in the regexp.")
   (defconst groovy-dollar-slashy-open-regex
     (rx "$/"))
   (defconst groovy-dollar-slashy-close-regex
-    (rx "/$")))
+    (rx "/$"))
+  (defconst groovy-declaration-keyword-regex
+    (rx
+     (* space)
```
Comment (Wilfred) [inline groovy-mode.el]: I suggest we use `symbol-start` rather than `space` to separate keywords. It's more robust.

File: groovy-mode.el [hunk 2]
```diff
@@ -147,29 +123,23 @@ The function name is the second group in the regexp.")
...
   "Matches interface names in groovy code.")
 
+(defvar groovy-annotation-regexp
+  (rx "@" symbol-start (+ (or (syntax word) (syntax symbol))) symbol-end)
+  "Match annotation names.")
+
+
 (defvar groovy-imenu-regexp
-  (list (list "Functions" groovy-function-regexp 2)
+  (list ;;(list "Functions" groovy-function-regexp 2)
```
Comment (Wilfred) [inline groovy-mode.el]: Does this no longer work? Why have you commented it out?

Comment (vincekd) [reply groovy-mode.el]: The function regexp was replaced with a scanning function. I'm not sure if imenu supports that?

**Comment by Wilfred** [comment]
Fantastic, thank you :)

**Comment by glasser** [comment]
I've updated to a version that has this, and I'm now getting these errors when I open a Gradle file:

```
Debugger entered--Lisp error: (void-variable groovy-declaration-regexp)
  groovy-declaration-search(2259)
  font-lock-fontify-keywords-region(1539 2259 nil)
  font-lock-default-fontify-region(1539 2259 nil)
  font-lock-fontify-region(1539 2259 nil)
  smerge-mode(1)
  smerge-start-session()
  vc-git-find-file-hook()
  apply(vc-git-find-file-hook nil)
  vc-call-backend(Git find-file-hook)
  vc-refresh-state()
  run-hooks(find-file-hook)
  after-find-file(nil t)
  find-file-noselect-1(#<buffer build.gradle> "~/Projects/Meteor/monorepo/go/build.gradle" nil nil "~/Projects/Meteor/monorepo/go/build.gradle" (8927342 16777220))
  find-file-noselect("/Users/glasser/Projects/Meteor/monorepo/go/build.gradle")
  (or (get-file-buffer file) (find-file-noselect file))
  (if (and (not force-worktree) (stringp rev)) (magit-find-file-noselect rev file) (or (get-file-buffer file) (find-file-noselect file)))
  (let* ((hunk (magit-diff-visit--hunk)) (line (and hunk (magit-diff-hunk-line hunk))) (col (and hunk (magit-diff-hunk-column hunk))) (rev (if (and magit-diff-visit-previous-blob (magit-section-match (quote hunk)) (save-excursion (goto-char (line-beginning-position)) (looking-at "-"))) (magit-diff-visit--range-beginning) (magit-diff-visit--range-end))) (buf (if (and (not force-worktree) (stringp rev)) (magit-find-file-noselect rev file) (or (get-file-buffer file) (find-file-noselect file))))) (magit-display-file-buffer buf) (let ((save-selected-window--state (internal--before-with-selected-window (or (get-buffer-window buf (quote visible)) (error "File buffer is not visible"))))) (save-current-buffer (unwind-protect (progn (select-window (car save-selected-window--state) (quote norecord)) (if line (progn (setq line ...) (let ... ...))) (if (magit-anything-unmerged-p file) (progn (smerge-start-session))) (run-hooks (quote magit-diff-visit-file-hook))) (internal--after-with-selected-window save-selected-window--state)))))
  (if (magit-file-accessible-directory-p file) (magit-diff-visit-directory file other-window) (let* ((hunk (magit-diff-visit--hunk)) (line (and hunk (magit-diff-hunk-line hunk))) (col (and hunk (magit-diff-hunk-column hunk))) (rev (if (and magit-diff-visit-previous-blob (magit-section-match (quote hunk)) (save-excursion (goto-char ...) (looking-at "-"))) (magit-diff-visit--range-beginning) (magit-diff-visit--range-end))) (buf (if (and (not force-worktree) (stringp rev)) (magit-find-file-noselect rev file) (or (get-file-buffer file) (find-file-noselect file))))) (magit-display-file-buffer buf) (let ((save-selected-window--state (internal--before-with-selected-window (or (get-buffer-window buf ...) (error "File buffer is not visible"))))) (save-current-buffer (unwind-protect (progn (select-window (car save-selected-window--state) (quote norecord)) (if line (progn ... ...)) (if (magit-anything-unmerged-p file) (progn ...)) (run-hooks (quote magit-diff-visit-file-hook))) (internal--after-with-selected-window save-selected-window--state))))))
  magit-diff-visit-file("/Users/glasser/Projects/Meteor/monorepo/go/build.gradle" nil)
  funcall-interactively(magit-diff-visit-file "/Users/glasser/Projects/Meteor/monorepo/go/build.gradle" nil)
  #<subr call-interactively>(magit-diff-visit-file nil nil)
  apply(#<subr call-interactively> magit-diff-visit-file (nil nil))
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> magit-diff-visit-file nil nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (magit-diff-visit-file nil nil))
  call-interactively(magit-diff-visit-file nil nil)
  command-execute(magit-diff-visit-file)
```

I'm not really sure how eval-when-compile works but I guess it isn't for me? I am loading groovy-mode via melpa.

**Pull request merged**